File: heap(priority queue)/kth_largest_element_in_array.py
from typing import List
import heapq


class Solution:
    # heap(priority queue)
    # tc O(nlogk), sc O(1)
    def findKthLargest1(self, nums: List[int], k: int) -> int:
        minHeap = []
        for num in nums:
            heapq.heappush(minHeap, num)
            if len(minHeap) > k:
                heapq.heappop(minHeap)

        return heapq.heappop(minHeap)

    # sorting
    # tc O(nlogn), sc O(1)
    def findKthLargest2(self, nums: List[int], k: int) -> int:
        nums.sort()
        return nums[len(nums) - k]

    def findKthLargest3(self, nums: List[int], k: int) -> int:
        minHeap = nums
        heapq.heapify(minHeap)
        return heapq.nlargest(k, minHeap)[-1]

File: heap(priority queue)/kth_largest_element_in_array.java
import java.util.*;

public class kth_largest_element_in_array {
    public static void main(String[] args) {
        
    }
}

class Solution {
    // heap(priority queue)
    // tc O(nlogk), sc O(1)
    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        for (int num : nums) {
            minHeap.offer(num);
            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }
        return minHeap.poll();
    }

    // sorting
    // tc O(nlogn), sc O(1)
    public int findKthLargest2(int[] nums, int k) {
        Arrays.sort(nums);
        return nums[nums.length - k];
    }
}